collector/pressure_linux.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nopressure
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/procfs"
+)
+
+var (
+	psiResources = []string{"cpu", "io", "memory"}
+	psiErrPrefix = "pressure information is not available"
+)
+
+type pressureStatsCollector struct {
+	cpu     *prometheus.Desc
+	io      *prometheus.Desc
+	ioFull  *prometheus.Desc
+	mem     *prometheus.Desc
+	memFull *prometheus.Desc
+}
+
+func init() {
+	registerCollector("pressure", defaultEnabled, NewPressureStatsCollector)
+}
+
+// NewPressureStatsCollector returns a Collector exposing pressure stall information
+func NewPressureStatsCollector() (Collector, error) {
+	return &pressureStatsCollector{
+		cpu: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "cpu_waiting_seconds_total"),
+			"Total time in seconds that processes have waited for CPU time",
+			nil, nil,
+		),
+		io: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "io_waiting_seconds_total"),
+			"Total time in seconds that processes have waited due to IO congestion",
+			nil, nil,
+		),
+		ioFull: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "io_stalled_seconds_total"),
+			"Total time in seconds no process could make progress due to IO congestion",
+			nil, nil,
+		),
+		mem: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "memory_waiting_seconds_total"),
+			"Total time in seconds that processes have waited for memory",
+			nil, nil,
+		),
+		memFull: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "memory_stalled_seconds_total"),
+			"Total time in seconds no process could make progress due to memory congestion",
+			nil, nil,
+		),
+	}, nil
+}
+
+// Update calls procfs.NewPSIStatsForResource for the different resources and updates the values
+func (c *pressureStatsCollector) Update(ch chan<- prometheus.Metric) error {
+	fs, err := procfs.NewFS(*procPath)
```

Comment:
Take a look at nfsd_linux.go, we put the `procfs.NewFS` stuff in the "New" function, rather than declare it for every Update().

Reply:
If I move this logic into `NewPressureStatsCollector` then it dies on startup:

```
FATA[0000] Couldn't create metrics handler: couldn't create collector: Not collecting pressure stall information, requires a kernel with CONFIG_PSI  source="node_exporter.go:55"
```

So I can move it there but I can't have it return an error, at least not with `defaultEnabled`?